File: tests/end-to-end/batch/java/src/main/java/com/myorg/Stack.java
package com.myorg;

import software.constructs.Construct;

import java.util.*;
import software.amazon.awscdk.CfnMapping;
import software.amazon.awscdk.CfnTag;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;

import software.amazon.awscdk.*;
import software.amazon.awscdk.services.batch.*;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.iam.*;

class BatchStack extends Stack {
    private Object computeEnvironmentArn;

    private Object jobQueueArn;

    private Object jobDefinitionArn;

    public Object getComputeEnvironmentArn() {
        return this.computeEnvironmentArn;
    }

    public Object getJobQueueArn() {
        return this.jobQueueArn;
    }

    public Object getJobDefinitionArn() {
        return this.jobDefinitionArn;
    }

    public BatchStack(final Construct scope, final String id) {
        super(scope, id, null);
    }

    public BatchStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        CfnRole batchServiceRole = CfnRole.Builder.create(this, "BatchServiceRole")
                .assumeRolePolicyDocument(Map.of("Version", "2012-10-17",
                "Statement", Arrays.asList(
                        Map.of("Effect", "Allow",
                        "Principal", Map.of("Service", "batch.amazonaws.com"),
                        "Action", "sts:AssumeRole"))))
                .managedPolicyArns(Arrays.asList(
                        "arn:aws:iam::aws:policy/service-role/AWSBatchServiceRole"))
                .build();

        CfnRole ecsInstanceRole = CfnRole.Builder.create(this, "EcsInstanceRole")
                .assumeRolePolicyDocument(Map.of("Version", "2008-10-17",
                "Statement", Arrays.asList(
                        Map.of("Sid", "",
                        "Effect", "Allow",
                        "Principal", Map.of("Service", "ec2.amazonaws.com"),
                        "Action", "sts:AssumeRole"))))
                .managedPolicyArns(Arrays.asList(
                        "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role"))
                .build();

        CfnInternetGateway internetGateway = CfnInternetGateway.Builder.create(this, "InternetGateway")
                .build();

        CfnJobDefinition jobDefinition = CfnJobDefinition.Builder.create(this, "JobDefinition")
                .type("container")
                .containerProperties(CfnJobDefinition.ContainerPropertiesProperty.builder()
                        .image(String.join("",
                                "137112412989.dkr.ecr.",
                                this.getRegion(),
                                ".amazonaws.com/amazonlinux:latest"))
                        .vcpus(2)
                        .memory(2000)
                        .command(Arrays.asList(
                                "echo",
                                "Hello world"))
                        .build())
                .retryStrategy(CfnJobDefinition.RetryStrategyProperty.builder()
                        .attempts(1)
                        .build())
                .build();

        CfnVPC vpc = CfnVPC.Builder.create(this, "VPC")
                .cidrBlock("10.0.0.0/16")
                .build();

        CfnInstanceProfile iamInstanceProfile = CfnInstanceProfile.Builder.create(this, "IamInstanceProfile")
                .roles(Arrays.asList(
                        ecsInstanceRole.getRef()))
                .build();

        CfnRouteTable routeTable = CfnRouteTable.Builder.create(this, "RouteTable")
                .vpcId(vpc.getRef())
                .build();

        CfnSecurityGroup securityGroup = CfnSecurityGroup.Builder.create(this, "SecurityGroup")
                .groupDescription("EC2 Security Group for instances launched in the VPC by Batch")
                .vpcId(vpc.getRef())
                .build();

        CfnSubnet subnet = CfnSubnet.Builder.create(this, "Subnet")
                .cidrBlock("10.0.0.0/24")
                .vpcId(vpc.getRef())
                .mapPublicIpOnLaunch(true)
                .build();

        CfnVPCGatewayAttachment vpcGatewayAttachment = CfnVPCGatewayAttachment.Builder.create(this, "VPCGatewayAttachment")
                .vpcId(vpc.getRef())
                .internetGatewayId(internetGateway.getRef())
                .build();

        CfnComputeEnvironment computeEnvironment = CfnComputeEnvironment.Builder.create(this, "ComputeEnvironment")
                .type("MANAGED")
                .computeResources(CfnComputeEnvironment.ComputeResourcesProperty.builder()
                        .type("EC2")
                        .minvCpus(0)
                        .desiredvCpus(0)
                        .maxvCpus(64)
                        .instanceTypes(Arrays.asList(
                                "optimal"))
                        .subnets(Arrays.asList(
                                subnet.getRef()))
                        .securityGroupIds(Arrays.asList(
                                securityGroup.getRef()))
                        .instanceRole(iamInstanceProfile.getRef())
                        .build())
                .serviceRole(batchServiceRole.getRef())
                .build();

        CfnRoute route = CfnRoute.Builder.create(this, "Route")
                .routeTableId(routeTable.getRef())
                .destinationCidrBlock("0.0.0.0/0")
                .gatewayId(internetGateway.getRef())
                .build();

        CfnSubnetRouteTableAssociation subnetRouteTableAssociation = CfnSubnetRouteTableAssociation.Builder.create(this, "SubnetRouteTableAssociation")
                .routeTableId(routeTable.getRef())
                .subnetId(subnet.getRef())
                .build();

        CfnJobQueue jobQueue = CfnJobQueue.Builder.create(this, "JobQueue")
                .priority(1)
                .computeEnvironmentOrder(Arrays.asList(
                        CfnJobQueue.ComputeEnvironmentOrderProperty.builder()
                                .order(1)
                                .computeEnvironment(computeEnvironment.getRef())
                                .build()))
                .build();

        this.computeEnvironmentArn = computeEnvironment.getRef();
        CfnOutput.Builder.create(this, "CfnOutputComputeEnvironmentArn")
                .key("ComputeEnvironmentArn")
                .value(this.computeEnvironmentArn.toString())
                .build();

        this.jobQueueArn = jobQueue.getRef();
        CfnOutput.Builder.create(this, "CfnOutputJobQueueArn")
                .key("JobQueueArn")
                .value(this.jobQueueArn.toString())
                .build();

        this.jobDefinitionArn = jobDefinition.getRef();
        CfnOutput.Builder.create(this, "CfnOutputJobDefinitionArn")
                .key("JobDefinitionArn")
                .value(this.jobDefinitionArn.toString())
                .build();

    }
}

File: tests/end-to-end/batch/csharp/Stack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.Batch;
using Amazon.CDK.AWS.EC2;
using Amazon.CDK.AWS.IAM;
using Constructs;
using System.Collections.Generic;

namespace BatchStack
{
    public class BatchStackProps : StackProps
    {
        public double MaxCpus { get; set; }

    }

    /// <summary>
    /// AWS CloudFormation Sample Template Managed Single Batch Job Queue: This template demonstrates the usage of simple Job Queue and EC2 style Compute Environment.  **WARNING** You will be billed for the AWS resources used if you create a stack from this template.
    /// </summary>
    public class BatchStack : Stack
    {
        public object ComputeEnvironmentArn { get; } 

        public object JobQueueArn { get; } 

        public object JobDefinitionArn { get; } 

        public BatchStack(Construct scope, string id, BatchStackProps props = null) : base(scope, id, props)
        {
            // Applying default props
            props ??= new BatchStackProps();
            props.MaxCpus = new CfnParameter(this, "MaxCpus", new CfnParameterProps
            {
                Type = "Number",
                Default = props.MaxCpus.ToString() ?? "64",
                NoEcho = true,
            }).ValueAsNumber;


            // Resources
            var batchServiceRole = new CfnRole(this, "BatchServiceRole", new CfnRoleProps
            {
                AssumeRolePolicyDocument = new Dictionary<string, object>
                {
                    { "Version", "2012-10-17"},
                    { "Statement", new []
                    {
                        new Dictionary<string, object>
                        {
                            { "Effect", "Allow"},
                            { "Principal", new Dictionary<string, object>
                            {
                                { "Service", "batch.amazonaws.com"},
                            }},
                            { "Action", "sts:AssumeRole"},
                        },
                    }},
                },
                ManagedPolicyArns = new []
                {
                    "arn:aws:iam::aws:policy/service-role/AWSBatchServiceRole",
                },
            });
            var ecsInstanceRole = new CfnRole(this, "EcsInstanceRole", new CfnRoleProps
            {
                AssumeRolePolicyDocument = new Dictionary<string, object>
                {
                    { "Version", "2008-10-17"},
                    { "Statement", new []
                    {
                        new Dictionary<string, object>
                        {
                            { "Sid", ""},
                            { "Effect", "Allow"},
                            { "Principal", new Dictionary<string, object>
                            {
                                { "Service", "ec2.amazonaws.com"},
                            }},
                            { "Action", "sts:AssumeRole"},
                        },
                    }},
                },
                ManagedPolicyArns = new []
                {
                    "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role",
                },
            });
            var internetGateway = new CfnInternetGateway(this, "InternetGateway", new CfnInternetGatewayProps
            {
            });
            var jobDefinition = new CfnJobDefinition(this, "JobDefinition", new CfnJobDefinitionProps
            {
                Type = "container",
                ContainerProperties = new CfnJobDefinition.ContainerPropertiesProperty
                {
                    Image = string.Join("", new []
                    {
                        "137112412989.dkr.ecr.",
                        Region,
                        ".amazonaws.com/amazonlinux:latest",
                    }),
                    Vcpus = 2,
                    Memory = 2000,
                    Command = new []
                    {
                        "echo",
                        "Hello world",
                    },
                },
                RetryStrategy = new CfnJobDefinition.RetryStrategyProperty
                {
                    Attempts = 1,
                },
            });
            var vpc = new CfnVPC(this, "VPC", new CfnVPCProps
            {
                CidrBlock = "10.0.0.0/16",
            });
            var iamInstanceProfile = new CfnInstanceProfile(this, "IamInstanceProfile", new CfnInstanceProfileProps
            {
                Roles = new []
                {
                    ecsInstanceRole.Ref,
                },
            });
            var routeTable = new CfnRouteTable(this, "RouteTable", new CfnRouteTableProps
            {
                VpcId = vpc.Ref,
            });
            var securityGroup = new CfnSecurityGroup(this, "SecurityGroup", new CfnSecurityGroupProps
            {
                GroupDescription = "EC2 Security Group for instances launched in the VPC by Batch",
                VpcId = vpc.Ref,
            });
            var subnet = new CfnSubnet(this, "Subnet", new CfnSubnetProps
            {
                CidrBlock = "10.0.0.0/24",
                VpcId = vpc.Ref,
                MapPublicIpOnLaunch = true,
            });
            var vpcGatewayAttachment = new CfnVPCGatewayAttachment(this, "VPCGatewayAttachment", new CfnVPCGatewayAttachmentProps
            {
                VpcId = vpc.Ref,
                InternetGatewayId = internetGateway.Ref,
            });
            var computeEnvironment = new CfnComputeEnvironment(this, "ComputeEnvironment", new CfnComputeEnvironmentProps
            {
                Type = "MANAGED",
                ComputeResources = new CfnComputeEnvironment.ComputeResourcesProperty
                {
                    Type = "EC2",
                    MinvCpus = 0,
                    DesiredvCpus = 0,
                    MaxvCpus = props.MaxCpus,
                    InstanceTypes = new []
                    {
                        "optimal",
                    },
                    Subnets = new []
                    {
                        subnet.Ref,
                    },
                    SecurityGroupIds = new []
                    {
                        securityGroup.Ref,
                    },
                    InstanceRole = iamInstanceProfile.Ref,
                },
                ServiceRole = batchServiceRole.Ref,
            });
            var route = new CfnRoute(this, "Route", new CfnRouteProps
            {
                RouteTableId = routeTable.Ref,
                DestinationCidrBlock = "0.0.0.0/0",
                GatewayId = internetGateway.Ref,
            });
            var subnetRouteTableAssociation = new CfnSubnetRouteTableAssociation(this, "SubnetRouteTableAssociation", new CfnSubnetRouteTableAssociationProps
            {
                RouteTableId = routeTable.Ref,
                SubnetId = subnet.Ref,
            });
            var jobQueue = new CfnJobQueue(this, "JobQueue", new CfnJobQueueProps
            {
                Priority = 1,
                ComputeEnvironmentOrder = new []
                {
                    new CfnJobQueue.ComputeEnvironmentOrderProperty
                    {
                        Order = 1,
                        ComputeEnvironment = computeEnvironment.Ref,
                    },
                },
            });

            // Outputs
            ComputeEnvironmentArn = computeEnvironment.Ref;
            JobQueueArn = jobQueue.Ref;
            JobDefinitionArn = jobDefinition.Ref;
        }
    }
}
